src/Planner/PlannerJoinTree.cpp

```diff
@@ -143,6 +151,106 @@ NameAndTypePair chooseSmallestColumnToReadFromStorage(const StoragePtr & storage
     return result;
 }
 
+bool applyTrivialCountIfPossible(
+    QueryPlan & query_plan,
+    const TableNode & table_node,
+    const QueryTreeNodePtr & query_tree,
+    const ContextPtr & query_context,
+    const Names & columns_names)
+{
+    const auto & settings = query_context->getSettingsRef();
+    if (!settings.optimize_trivial_count_query)
+        return false;
+
+    /// can't apply if FINAL
+    if (table_node.getTableExpressionModifiers().has_value() && table_node.getTableExpressionModifiers()->hasFinal())
+        return false;
+
+    auto & main_query_node = query_tree->as<QueryNode &>();
+    if (main_query_node.hasGroupBy())
+        return false;
+
+    const auto & storage = table_node.getStorage();
+    if (!storage || storage->hasLightweightDeletedMask())
+        return false;
+
+    if (settings.max_parallel_replicas > 1 || //
+        settings.allow_experimental_query_deduplication || //
+        settings.empty_result_for_aggregation_by_empty_set)
+        return false;
+
+    QueryTreeNodes aggregates = collectAggregateFunctionNodes(query_tree);
+    if (aggregates.size() != 1)
+        return false;
+
+    const auto * function_node = typeid_cast<const FunctionNode *>(aggregates.front().get());
+    if (!function_node)
+        return false;
+
+    if (!function_node->getAggregateFunction())
+        return false;
+
+    const auto * count_func = typeid_cast<const AggregateFunctionCount *>(function_node->getAggregateFunction().get());
+    if (!count_func)
+        return false;
+
+    /// get number of rows
+    std::optional<UInt64> num_rows{};
+    /// Transaction check here is necessary because
+    /// MergeTree maintains total count for all parts in Active state and it simply returns that number for trivial select count() from table query.
+    /// But if we have current transaction, then we should return number of rows in current snapshot (that may include parts in Outdated state),
+    /// so we have to use totalRowsByPartitionPredicate() instead of totalRows even for trivial query
+    /// See https://github.com/ClickHouse/ClickHouse/pull/24258/files#r828182031
+    if (!main_query_node.hasPrewhere() && !main_query_node.hasWhere() && !query_context->getCurrentTransaction())
```

Comment:
Consider to move this check before we collect aggregate functions, because they are easier to perform.

Reply:
Okay. Lets see how it will look like after partition predicate optimization.